core/wasm-interface/src/lib.rs

```diff
@@ -0,0 +1,285 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Types and traits for interfacing between the host and the wasm runtime.
+
+use std::{borrow::Cow, marker::PhantomData, mem, iter::Iterator};
+
+mod wasmi_impl;
+
+/// Value types supported by Substrate on the boundary between host/Wasm.
+#[derive(Copy, Clone, PartialEq, Debug, Eq)]
+pub enum ValueType {
+	/// An `i32` value type.
+	I32,
+	/// An `i64` value type.
+	I64,
+}
+
+/// Values supported by Substrate on the boundary between host/Wasm.
+#[derive(Eq, PartialEq, Debug, Clone, Copy)]
+pub enum Value {
+	/// An `i32` value.
+	I32(i32),
+	/// An `i64` value.
+	I64(i64),
+}
+
+/// Type to represent a pointer in wasm at the host.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct Pointer<T> {
+	ptr: u32,
+	_marker: PhantomData<T>,
+}
+
+impl<T> Pointer<T> {
+	/// Create a new instance of `Self`.
+	pub fn new(ptr: u32) -> Self {
+		Self {
+			ptr,
+			_marker: Default::default(),
+		}
+	}
+
+	/// Calculate the offset from this pointer.
+	///
+	/// `offset` is in units of `T`. So, `3` means `3 * mem::size_of::<T>()` as offset to the pointer.
+	pub fn offset(self, offset: u32) -> Self {
```

Comment:
This function seems to be a bit of a footgun: `size_of` is the size of of a type on the host machine while this type is supposed to be used in the wasm context.

Reply:
Before we have done this manually. Working with pointers is always a footgun :D We could enforce that the user just can give a `ValueType`.